### ISSUE TITLE
Add support for using unit information from dataflash logs

### ISSUE THREAD

**Issue by peterbarker** [ArduPilot/apm_planner#1089]
ArduPilot's dataflash logs now contain unit information for each and every field.

There are several new messages involved in this - but they act in the same way as `FMT` messages, being basic "header information".

The new `UNIT` message defines a mapping between a character and a unit.  These units are, for the most part, SI units - `volts`, `seconds`, `Hz` - and derived units such as `m/s`, however there are some "special case" units like `satellites` and `deglatitude`.  It is expected that a GCS be able to convert these into units presentable to a user (perhaps according to their preference), so for example temperature in Fahrenheit in Liberia.

The new `MULT` message defines a mapping mapping between a character and a multiplier.  So, for example, `2` indicates a multiplier of `100`.  These are used where the natural presentation of a time series in a log is not in the base SI unit - the obvious example being `TimeUS` - a time in microseconds since boot.

The new `FMTU` message provides the unit for each field present in the corresponding FMT message.  `FmtType` gives the `FMT` number, and the other two fields present both the unit and multiplier for the corresponding (by offset) field in the `FMT` message.  For example, almost every message we emit starts with a `TimeUS` field, a time since system boot in microseconds.  The `FMTU` message for each of these would have `s` in `UnitIds[0]` and `F` in `MultIds` (i.e. `10^-6`).  `F` is a reference into the table formed by `MULT` and `s` is a reference into the table formed by `UNIT`

```
2018-01-24 10:13:34.20: FMTU {TimeUS : 1687658, FmtType : 229, UnitIds : s--, MultIds : F--}
```

Possible uses include presenting fields in locale-specific forms, labelling graph axes, and possibly determining which time series might be presented on the same axis (given they have the same unit and a similar multiplier).  In tools where formulae can be used to present derived values it would be possible to enforce unit-correctness - only subtracting seconds from seconds and metres from metres, for example, and where the language permits adding units generally to values it would be possible to enforce comparisons only against like-units.

**Comment by peterbarker** [comment]
@Arne-W Only the FMT units are guaranteed to be written at the moment.  Sorry.

**Comment by Arne-W** [comment]
@peterbarker Ok - not a very big problem :wink: 
One question left was a little bit too optimistic :smiley: here is the second: I have a 3.6-dev log from a user for testing the NaN feature. The first look showed that there is a `UNIT` entry for `- (45)` which is empty and a `UNIT` for `? (63)` which is unknown but no `MULT` entry although they are used in the MultIds field of the `FMTU` messages. Is it a bug or did I miss something??

**Comment by peterbarker** [comment]
@Arne-W `-` is "no applicable unit".  These are usually "I'm too lazy to work out what the correct unit is" - like the PID values.

`?` is "there is probably an applicable unit here, but I'm too lazy to work out what that is".

I'm responsible for the vast majority of these...

**Comment by Arne-W** [comment]
@peterbarker Yes I got that - but those (`-` and `?`) are used in the MultIds field as well but there is no matching `MULT` message. I thought there should be an `MULT` message with ID 45 and one with ID 63 as there are those IDs in the `UNIT` messages.

![missingmult](https://user-images.githubusercontent.com/7621911/35362047-057ff3e2-0164-11e8-9c3c-f99ca313f84b.jpeg)
No `MULT` message for ID 45 or 63.

![unit](https://user-images.githubusercontent.com/7621911/35362151-6c9ef56e-0164-11e8-85cc-8b3e0e7a5e08.jpeg)
`UNIT` has those IDs

**Comment by Arne-W** [comment]
Just for completeness - see ArduPilot/ardupilot#7609 which adds the missing `MULT` messages.
Thanks @peterbarker for this incredible fast fix 👍

**Comment by Arne-W** [comment]
@peterbarker While implementing this feature I found some weird scaling. Especially the GPS messages.
![scaling](https://user-images.githubusercontent.com/7621911/35777134-e6f84b48-09a8-11e8-87bc-8cd3f0b96c6f.jpeg)
Here you can see the scaled content of such a message - look at the HDop, Lat and Lng colums they seem much too small?! The GPS message has the ID 130 the screen shots above show the FMTU content. Could you please check if its correct?!

**Comment by peterbarker** [comment]
@Arne-W Are you applying the multiplier for both the type (`L`) and the multiplier supplied in the `MULT` column?  If so, don't.  If using the muliplier column, use only the width (and sign) of the column type field.

So when using the `MULT` multiplier, `E` and `I` are equivalent in the type specification column.

https://github.com/ardupilot/ardupilot/blob/master/libraries/DataFlash/LogStructure.h#L106


In the far-flung future I would the format type to be pretty much a c-type.

**Comment by meee1** [comment]
there are 5 types that needs to not be double multiplied
cCeEL

**Comment by Arne-W** [comment]
Hmmm - so did I get you right - There are 2 concurrent scaling Informations in the log?? Will that be changed until the release of the new format?

**Comment by peterbarker** [comment]
On Mon, 5 Feb 2018, Arne Wischmann wrote:

> Hmmm - so did I get you right - There are 2 concurrent scaling Informations
> in the log?? Will that be changed until the release of the new format?

At some stage I would like to eliminate all of the types that infer a 
scaling.  But that can't be done until all tools that use that scaling 
information are aware that there may be explicit scaling information - 
which won't be for a while!  So if a tool does become aware that there's 
explicit scaling information it should only use the explicit scaling 
information.

**Comment by Arne-W** [comment]
@peterbarker Ah - yes ok backward compatibility is always a thing I like to forget - things are getting complicated if all the old stuff shall still work. But nevertheless I found a possibility to handle both scaling Informations.
And as always - thanks for the fast support 👍

**Comment by Arne-W** [comment]
This issue is fixed by 4bb3d410b99f22c39534c74b58e26ec20b8541d2 and e3357d026b3fc9c44044993e070752723f85113f
